lxd/auth/driver_openfga.go

```diff
@@ -0,0 +1,465 @@
+
+	"github.com/canonical/lxd/lxd/identity"
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/api"
+	"github.com/canonical/lxd/shared/entity"
+	"github.com/canonical/lxd/shared/logger"
+)
+
+//go:embed driver_openfga_model.openfga
+var model string
+
+// embeddedOpenFGA implements Authorizer using an embedded OpenFGA server.
+type embeddedOpenFGA struct {
+	commonAuthorizer
+	tlsAuthorizer *tls
+	server        openfgav1.OpenFGAServiceServer
+	identityCache *identity.Cache
+}
+
+// The OpenFGA server requires a ULID to specify the store that we are querying against.
+// Our storage.OpenFGADatastore implementation only has one store, so use a dummy value.
+var dummyDatastoreULID = ulid.Make().String()
+
+// load sets up the authorizer.
+func (e *embeddedOpenFGA) load(ctx context.Context, identityCache *identity.Cache, opts Opts) error {
+	if identityCache == nil {
+		return fmt.Errorf("Must provide certificate cache")
+	}
+
+	e.identityCache = identityCache
+
+	// Use the TLS driver for TLS authenticated users for now.
+	tlsDriver := &tls{}
+	err := tlsDriver.load(ctx, identityCache, opts)
+	if err != nil {
+		return err
+	}
+
+	e.tlsAuthorizer = tlsDriver
+
+	if opts.openfgaDatastore == nil {
+		return fmt.Errorf("The OpenFGA datastore option must be set")
+	}
+
+	openfgaServerOptions := []server.OpenFGAServiceV1Option{
+		// Use our embedded datastore.
+		server.WithDatastore(opts.openfgaDatastore),
+		// Use our logger.
+		server.WithLogger(openfgaLogger{l: e.logger}),
+		// Set the max concurrency to 1 for both read and check requests.
+		// Our driver cannot perform concurrent reads.
+		server.WithMaxConcurrentReadsForListObjects(1),
+		server.WithMaxConcurrentReadsForCheck(1),
+	}
+
+	e.server, err = server.NewServerWithOpts(openfgaServerOptions...)
+	if err != nil {
+		return err
+	}
+
+	// Transform the model from the DSL into the protobuf type.
+	protoModel, err := transformer.TransformDSLToProto(model)
+	if err != nil {
+		return err
+	}
+
+	// Write the model to the server.
+	_, err = e.server.WriteAuthorizationModel(ctx, &openfgav1.WriteAuthorizationModelRequest{
+		StoreId:         dummyDatastoreULID,
+		TypeDefinitions: protoModel.TypeDefinitions,
+		SchemaVersion:   protoModel.SchemaVersion,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CheckPermission checks whether the user who sent the request has the given entitlement on the given entity using the
+// embedded OpenFGA server.
+func (e *embeddedOpenFGA) CheckPermission(ctx context.Context, r *http.Request, entityURL *api.URL, entitlement Entitlement) error {
+	logCtx := logger.Ctx{"entity_url": entityURL.String(), "entitlement": entitlement, "request_url": r.URL.String(), "method": r.Method}
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	// Inspect request.
+	details, err := e.requestDetails(r)
+	if err != nil {
+		return api.StatusErrorf(http.StatusForbidden, "Failed to extract request details: %v", err)
```

Comment:
Should be %w

Reply:
Actually in this case I'm not sure it can be? Do we need `api.StatusError` to implement `Unwrap` for this to work properly?

Reply:
ah yeah ofcourse you're correct. Although now im looking closer how do we know that the err means forbidden anyway?

Reply:
`api.StatusErrorCheck` is able to unwrap until it gets to the `api.StatusError` and then it can check the status code. I've just implemented the interface for it now so we can pass in `%w` and wrap underlying errors. Should mean we will be able to distinguish a general 404 into `sql.ErrNoRows` or `io.ErrNotExist` in future.

Reply:
OK, but what I mean is that this actually feels like the wrong place to be returning api.StatusErrorf(http.StatusForbidden), shouldn't that be done in e.requestDetails so it has the option of returning some other sort of status code (internal server error for example or bad request) if the issue a bad request or an internal issue?

Reply:
Ahhh I see what you mean. Yes I will change it.

Reply:
can include in https://github.com/canonical/lxd/pull/13077